storm-server/src/main/java/org/apache/storm/daemon/nimbus/HeartbeatCache.java

```diff
@@ -0,0 +1,229 @@
+        Map<List<Integer>, ExecutorCache> topoCache = cache.computeIfAbsent(topoId, MAKE_MAP);
+        if (executorBeats == null) {
+            executorBeats = new HashMap<>();
+        }
+
+        for (List<Integer> executor : allExecutors) {
+            final Map<String, Object> newBeat = executorBeats.get(executor);
+            ExecutorCache currBeat = topoCache.computeIfAbsent(executor, (k) -> new ExecutorCache(newBeat));
+            currBeat.updateFromHb(timeout, newBeat);
+        }
+    }
+
+    /**
+     * Update the heartbeats for a given worker.
+     * @param workerHeartbeat the heartbeats from the worker.
+     * @param taskTimeoutSecs the timeout we should be looking at.
+     */
+    public void updateHeartbeat(SupervisorWorkerHeartbeat workerHeartbeat, Integer taskTimeoutSecs) {
+        Map<List<Integer>, Map<String, Object>> executorBeats = StatsUtil.convertWorkerBeats(workerHeartbeat);
+        String topoId = workerHeartbeat.get_storm_id();
+        Map<List<Integer>, ExecutorCache> topoCache = cache.computeIfAbsent(topoId, MAKE_MAP);
+
+        for (ExecutorInfo executorInfo : workerHeartbeat.get_executors()) {
+            List<Integer> executor = Arrays.asList(executorInfo.get_task_start(), executorInfo.get_task_end());
+            final Map<String, Object> newBeat = executorBeats.get(executor);
+            ExecutorCache currBeat = topoCache.computeIfAbsent(executor, (k) -> new ExecutorCache(newBeat));
+            currBeat.updateFromHb(taskTimeoutSecs, newBeat);
+        }
+    }
+
+    /**
+     * Get all of the alive executors for a given topology.
+     * @param topoId the id of the topology we are looking for.
+     * @param allExecutors all of the executors for this topology.
+     * @param assignment the current topology assignment.
+     * @param taskLaunchSecs timeout for right after a worker is launched.
+     * @return the set of tasks that are alive.
+     */
+    public Set<List<Integer>> getAliveExecutors(String topoId, Set<List<Integer>> allExecutors, Assignment assignment, int taskLaunchSecs) {
+        Map<List<Integer>, ExecutorCache> topoCache = cache.computeIfAbsent(topoId, MAKE_MAP);
+        LOG.debug("Computing alive executors for {}\nExecutors: {}\nAssignment: {}\nHeartbeat cache: {}",
+            topoId, allExecutors, assignment, topoCache);
+
+        Set<List<Integer>> ret = new HashSet<>();
+        Map<List<Long>, Long> execToStartTimes = assignment.get_executor_start_time_secs();
+
+        for (List<Integer> exec : allExecutors) {
+            List<Long> longExec = new ArrayList<>(exec.size());
+            for (Integer num : exec) {
+                longExec.add(num.longValue());
+            }
+
+            Long startTime = execToStartTimes.get(longExec);
+            ExecutorCache executorCache = topoCache.get(exec);
+            //null isTimedOut means worker never reported any heartbeat
+            Boolean isTimedOut = executorCache == null ? null : executorCache.isTimedOut();
```

Comment:
Nit: Would probably be simpler to set isTimedOut to false if executorCache is null. That avoids the extra null check below.

Reply:
Yes I agree.  Will do it.  Again this was code that was already there, and I just refactored it, but I am happy to fix it.